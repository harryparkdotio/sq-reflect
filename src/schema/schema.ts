import * as pg from 'pg';

import { EnumDefinition, ClassDefinition } from './definitions';

export const getEnums = async (
  client: pg.Client,
  schema: string = 'public'
): Promise<EnumDefinition[]> => {
  const queryText = `
    SELECT
      t.oid::INT "id",
      t.typname "name",
      n.oid::INT "namespace_id",
      n.nspname "namespace",
      ta.oid::INT "arr_id",
      ta.typname "arr_name",
      enum.values
    FROM (
      SELECT
        e.enumtypid::INT "id",
        array_agg((
          SELECT e.enumlabel::TEXT
          ORDER BY e.enumsortorder ASC
        )) "values"
      FROM pg_catalog.pg_enum e
      GROUP BY e.enumtypid
    ) "enum"
    JOIN pg_catalog.pg_type t ON t.oid = enum.id
    JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
    LEFT JOIN pg_catalog.pg_type ta ON ta.oid = t.typarray AND t.typarray > 0
    WHERE
      n.nspname = $1
    ORDER BY t.typname ASC;
  `;

  return client.query(queryText, [schema]).then(r => r.rows);
};

export const getClasses = async (
  client: pg.Client,
  schema: string = 'public'
): Promise<ClassDefinition[]> => {
  const queryText = `
    SELECT
      classes.id,
      classes.name,
      classes.type,
      classes.namespace_id,
      classes.namespace,
      classes.attributes
    FROM (
      SELECT
        c.oid::INT "id",
        c.relname "name",
        c.relkind "type",
        n.oid::INT "namespace_id",
        n.nspname "namespace",
        json_agg(row_to_json(attr)) "attributes"
      FROM pg_catalog.pg_class c
      JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
      CROSS JOIN LATERAL (
        SELECT
          a.attname "name",
          NOT a.attnotnull "nullable",
          a.attnum "number",
          json_build_object(
            'id', t.oid::INT,
            'name', t.typname,
            'namespace_id', n.oid::INT,
            'namespace', n.nspname
          ) "type",
          pg_catalog.pg_get_expr(d.adbin, d.adrelid) "default"
        FROM pg_catalog.pg_attribute a
        JOIN pg_catalog.pg_type t ON t.oid = a.atttypid
        JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
        LEFT JOIN pg_catalog.pg_attrdef d ON d.adrelid = a.attrelid AND d.adnum = a.attnum AND a.atthasdef
        WHERE
          a.attrelid = c.oid
          AND a.attnum > 0
          AND NOT a.attisdropped
      ) "attr"
      WHERE
        -- relkind = 'r' denotes an ordinary table
        -- relkind = 'v' denotes a view
        c.relkind = ANY('{"r","v"}')
        AND n.nspname = $1
      GROUP BY
        c.oid,
        c.relname,
        c.relkind,
        n.oid,
        n.nspname
    ) "classes"
    ORDER BY classes.name ASC;
  `;

  return client.query(queryText, [schema]).then(r => r.rows);
};
