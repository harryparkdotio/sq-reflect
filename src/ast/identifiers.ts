export const reservedWords = [
  'abstract',
  'any',
  'await',
  'boolean',
  'break',
  'byte',
  'case',
  'catch',
  'char',
  'class',
  'const',
  'continue',
  'debugger',
  'default',
  'delete',
  'do',
  'double',
  'else',
  'enum',
  'export',
  'extends',
  'false',
  'final',
  'finally',
  'float',
  'for',
  'function',
  'goto',
  'if',
  'implements',
  'import',
  'in',
  'instanceof',
  'int',
  'interface',
  'let',
  'long',
  'native',
  'new',
  'null',
  'number',
  'object',
  'package',
  'private',
  'protected',
  'public',
  'return',
  'short',
  'static',
  'string',
  'super',
  'switch',
  'symbol',
  'synchronized',
  'this',
  'throw',
  'throws',
  'transient',
  'true',
  'try',
  'typeof',
  'var',
  'void',
  'volatile',
  'while',
  'with',
  'yield',
];

// IdentifierName can be written as unquoted property names, but may be reserved words
export function isIdentifierName(s: string) {
  return /^[$A-Z_][0-9A-Z_$]*$/i.test(s);
}

// Identifiers are e.g. legal variable names. They may not be reserved words
export function isIdentifier(s: string) {
  return isIdentifierName(s) && reservedWords.indexOf(s) < 0;
}
