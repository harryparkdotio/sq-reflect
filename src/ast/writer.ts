import * as ts from 'typescript';

export const astWriter = (statements: ts.Statement[]): string[] => {
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
    removeComments: false,
  });

  const sourceFile = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS
  );

  return statements.map(statement =>
    printer.printNode(ts.EmitHint.Unspecified, statement, sourceFile)
  );
};
