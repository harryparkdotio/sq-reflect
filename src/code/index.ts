import { source } from 'common-tags';

import { EnumDefinition, TableDefinition, TypeDefinition } from '../sql/definitions';
import { Transform } from '../transform';

import { NamingStrategy, PassiveNamingStrategy } from './naming-strategy';

export type CodeOptions = Partial<{
  namingStrategy: NamingStrategy;
}>;

export class Code {
  private readonly ns: NamingStrategy;

  constructor(options: CodeOptions = {}) {
    this.ns = options.namingStrategy || new PassiveNamingStrategy();
  }

  enum(definition: EnumDefinition) {
    const _enum = Transform.reserved(this.ns.enum(definition.name));

    return source`
      export enum ${_enum} {
        ${definition.values.sort().map(val => `${val} = '${val}',`)}
      }
    `;
  }

  table(definition: TableDefinition) {
    const _namespace = Transform.reserved(this.ns.namespace(`${definition.name}_fields`));
    const _interface = Transform.reserved(this.ns.interface(definition.name));

    interface Cols {
      type: string;
      prop: string;
      ref: string;
    }

    const cols = definition.columns.map<Cols>(col => ({
      type: this.type(col.type),
      prop: this.ns.property(col.name),
      ref: Transform.reserved(this.ns.type(col.name)),
    }));

    return source`
      export namespace ${_namespace} {
        ${cols.map(col => `export type ${col.ref} = ${col.type};`)}
      }

      export interface ${_interface} {
        ${cols.map(col => `${col.prop}: ${_namespace}.${col.ref};`)}
      }
    `;
  }

  type(definition: TypeDefinition) {
    const types = [definition.type || 'any', definition.nullable && 'null'];

    return types.filter(type => type).join(' | ');
  }
}
