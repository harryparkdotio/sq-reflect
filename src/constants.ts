export const reservedKeywords: string[] = [
  'any',
  'boolean',
  'break',
  'case',
  'catch',
  'class',
  'const',
  'continue',
  'debugger',
  'default',
  'delete',
  'else',
  'enum',
  'extends',
  'false',
  'finally',
  'for',
  'function',
  'if',
  'implements',
  'in',
  'instanceof',
  'interface',
  'let',
  'new',
  'null',
  'number',
  'object',
  'package',
  'private',
  'protected',
  'public',
  'return',
  'static',
  'string',
  'super',
  'switch',
  'symbol',
  'this',
  'throw',
  'true',
  'try',
  'typeof',
  'var',
  'void',
  'while',
  'with',
  'yield',
];
