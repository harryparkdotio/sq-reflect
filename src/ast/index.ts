import * as ts from 'typescript';
import * as Case from 'change-case';

import { flatten } from '../utils';
import {
  ClassDefinition,
  AttributeDefinition,
  EnumDefinition,
} from '../schema/definitions';
import { OidMap, typeGetterFn, TypeGetterFn } from '../schema/types';

import { genericTypeTransformer } from './transformers/generic-type';
import { isIdentifier } from './identifiers';

export const Transform = {
  enumName: (text: string) => Case.pascal(text),
  enumMember: (text: string) => Case.constant(text),
  interfaceName: (text: string) => Case.pascal(text),
  interfaceTypeParameterName: (text: string) => Case.pascal(text),
  interfacePropertyName: (text: string) => Case.snake(text),
};

export const getTypeFromAttribute = (
  attribute: AttributeDefinition,
  typeGetter: TypeGetterFn
): [ts.TypeNode] | [ts.TypeNode, ts.TypeNode] => {
  const typeNode =
    typeGetter(attribute.type.id) ||
    ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword);

  if (attribute.nullable) {
    return [typeNode, ts.createNull()];
  }

  return [typeNode];
};

const getGenericType = (
  attributeName: string,
  typeNode: ts.TypeNode | ts.TypeNode[]
): [ts.TypeNode[], ts.TypeParameterDeclaration[]] => {
  const typeParameterReferenceArray: ts.TypeParameterDeclaration[] = [];
  const transformers: ts.TransformerFactory<ts.TypeNode>[] = [
    genericTypeTransformer(attributeName, typeParameterReferenceArray),
  ];

  return [
    ts.transform(typeNode, transformers).transformed,
    typeParameterReferenceArray,
  ];
};

/* istanbul ignore next */
const makeIdentifier = (s: string): ts.Identifier =>
  (isIdentifier(s) && ts.createIdentifier(s)) || makeIdentifier(s + '_');

export const buildEnum = (
  definition: EnumDefinition
): [ts.EnumDeclaration, OidMap] => {
  const enumDeclaration = ts.createEnumDeclaration(
    undefined,
    [ts.createToken(ts.SyntaxKind.ExportKeyword)],
    ts.createIdentifier(Transform.enumName(definition.name)),
    definition.values.map(value =>
      ts.createEnumMember(
        makeIdentifier(Transform.enumMember(value)),
        ts.createStringLiteral(value)
      )
    )
  );

  const enumTypeReference = ts.createTypeReferenceNode(
    enumDeclaration.name,
    undefined
  );

  const customOidMap: OidMap = { [definition.id]: enumTypeReference };

  if (definition.arr_id) {
    customOidMap[definition.arr_id] = ts.createArrayTypeNode(enumTypeReference);
  }

  return [enumDeclaration, customOidMap];
};

export const buildAttribute = (
  attribute: AttributeDefinition,
  typeGetter: TypeGetterFn
): [ts.TypeParameterDeclaration[], ts.PropertySignature] => {
  const attributeType = getTypeFromAttribute(attribute, typeGetter);

  const [genericType, genericTypeTypeParameters] = getGenericType(
    attribute.name,
    attributeType
  );

  const propertySignature = ts.createPropertySignature(
    undefined,
    ts.createIdentifier(Transform.interfacePropertyName(attribute.name)),
    undefined,
    ts.createUnionTypeNode(genericType),
    undefined
  );

  return [genericTypeTypeParameters, propertySignature];
};

export const buildClass = (
  _class: ClassDefinition,
  typeGetter: TypeGetterFn
): ts.InterfaceDeclaration => {
  const interfaceNameIdentifier = ts.createIdentifier(
    Transform.interfaceName(_class.name)
  );

  const attributes = _class.attributes.map(a => buildAttribute(a, typeGetter));

  const interfaceTypeParameters: ts.TypeParameterDeclaration[] = flatten(
    attributes.map(a => a[0])
  );
  const interfaceMembers: ts.TypeElement[] = attributes.map(a => a[1]);

  const interfaceDeclaration = ts.createInterfaceDeclaration(
    undefined,
    [ts.createToken(ts.SyntaxKind.ExportKeyword)],
    interfaceNameIdentifier,
    interfaceTypeParameters,
    undefined,
    interfaceMembers
  );

  return interfaceDeclaration;
};

export const buildAst = (
  classes: ClassDefinition[],
  enums: EnumDefinition[]
) => {
  const assembledEnums = enums.map(e => buildEnum(e));
  const enumTypeReferences = assembledEnums.map(e => e[1]);

  const assembledClasses = classes.map(c =>
    buildClass(c, typeGetterFn(...enumTypeReferences))
  );

  const statements: ts.Statement[] = [
    ...assembledEnums.map(e => e[0]),
    ...assembledClasses,
  ];

  return statements;
};
