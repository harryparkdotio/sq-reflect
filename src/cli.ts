import * as meow from 'meow';
import * as pg from 'pg';
import * as fs from 'fs';

import * as ast from './ast';
import * as schema from './schema/schema';
import { astWriter } from './ast/writer';

const cli = meow(
  `
    Usage
      $ sq-reflect --conn postgres://postgres@localhost:5432/db_name --file schema.ts
 
    Options
      --conn, -c  postgres connection string
      --file, -f  filename to save generated types
      --schema, -s  postgres schema
 
    Examples
      $ sq-reflect --conn <conn> --file schema.ts
`,
  {
    flags: {
      help: {
        alias: 'h',
      },
      version: {
        alias: 'v',
      },
      conn: {
        type: 'string',
        alias: 'c',
      },
      schema: {
        type: 'string',
        alias: 's',
        default: 'public',
      },
    },
  }
);

const run = async (cli: meow.Result) => {
  const client = new pg.Client(cli.flags.conn);

  await client.connect();

  const classes = await schema.getClasses(client, cli.flags.schema);
  const enums = await schema.getEnums(client, cli.flags.schema);

  await client.end();

  const statements = ast.buildAst(classes, enums);
  const strings = astWriter(statements).join('\n\n');

  fs.writeFileSync(cli.flags.file, strings);
};

run(cli).catch(err => {
  console.error(err);
});
