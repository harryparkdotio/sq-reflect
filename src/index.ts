import { source } from 'common-tags';
import * as fse from 'fs-extra';

import { Code } from './code';
import { CamelCaseNamingStrategy, NamingStrategy, SnakeCaseNamingStrategy } from './code/naming-strategy';
import { Driver } from './driver/interfaces';
import { Postgres } from './driver/postgres';
import { EnumDefinition, TableDefinition } from './sql/definitions';
import { Meta } from './sql/meta';

interface Options {
  conn: string;
  schema: string;
  filename: string;
  case: string;
  meta: boolean;
  generics: boolean;
}

interface Data {
  enums: EnumDefinition[];
  tables: TableDefinition[];
}

export class SqReflect {
  private db: Driver;
  private meta: Meta;
  private data: Data;
  public content?: string;
  private filename: string;
  private code: Code;

  constructor(options: Options) {
    this.db = new Postgres(options.conn);

    let namingStrategy: NamingStrategy;

    switch (options.case) {
      case 'camel':
        namingStrategy = new CamelCaseNamingStrategy();
        break;
      case 'snake':
      default:
        namingStrategy = new SnakeCaseNamingStrategy();
        break;
    }

    this.meta = new Meta(this.db, { schema: options.schema });

    this.filename = options.filename;

    this.code = new Code({
      namingStrategy,
      emitMetadata: options.meta,
      emitGenerics: options.generics,
    });
  }

  async scrape() {
    await this.db.connect();

    this.data = {
      enums: await this.meta.Enums(),
      tables: await this.meta.Tables(),
    };

    if (this.data.tables.length === 0) {
      // tslint:disable-next-line:no-console
      console.warn('no tables were found.');
    }

    await this.db.disconnect();
  }

  async compile() {
    const code = this.code;
    const { enums, tables } = this.data;

    // define custom types
    enums.map(e => code.define(e.name));

    this.content = source`
      /**
       * THIS IS AN AUTO GENERATED FILE - DO NOT EDIT
       */

      /* eslint-disable */
      /* tslint:disable */

      ${[...enums.map(e => code.enum(e)), ...tables.map(t => code.table(t))].join('\n\n')}
    `;
  }

  async save() {
    await fse.ensureFile(this.filename);
    await fse.writeFile(this.filename, this.content);
  }
}
