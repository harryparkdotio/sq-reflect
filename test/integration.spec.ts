import { client } from './client';
import * as schema from './../src/schema/schema';
import * as ast from './../src/ast';
import { astWriter } from './../src/ast/writer';

describe('integration', () => {
  beforeEach(async () => {
    await client.query(`
      CREATE TYPE enum_abc_color AS ENUM ('PENDING', 'COMPLETED', 'STARTED', 'READY', 'DONE', 'STOPPED', 'PAUSED', 'PARTIALLY_COMPLETE');
      CREATE TYPE enum_def_status AS ENUM ('BLUE', 'RED', 'GREEN', 'ORANGE', 'YELLOW');

      CREATE VIEW "some_data_2" AS (
        SELECT
          1 "a_number",
          'a' "a_letter",
          '{}'::JSON "an_object"
      );

      CREATE TABLE "not_user" (
        id         UUID PRIMARY KEY,
        metadata   JSON NOT NULL,
        def_status enum_def_status NULL,
        created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
        deleted_at TIMESTAMPTZ NULL
      );
    `);
  });

  afterEach(async () => {
    await client.query(`
      DROP TABLE IF EXISTS "not_user";
      DROP VIEW IF EXISTS "some_data_2";

      DROP TYPE IF EXISTS enum_abc_color;
      DROP TYPE IF EXISTS enum_def_status;
    `);
  });

  it('should return type declarations', async () => {
    const classes = await schema.getClasses(client, 'public');
    const enums = await schema.getEnums(client, 'public');

    const statements = ast.buildAst(classes, enums);
    expect(astWriter(statements).join('\n\n')).toMatchSnapshot();
  });
});
