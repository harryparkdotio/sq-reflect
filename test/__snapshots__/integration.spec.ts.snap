// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration should return type declarations 1`] = `
"export enum EnumAbcColor {
    PENDING = \\"PENDING\\",
    COMPLETED = \\"COMPLETED\\",
    STARTED = \\"STARTED\\",
    READY = \\"READY\\",
    DONE = \\"DONE\\",
    STOPPED = \\"STOPPED\\",
    PAUSED = \\"PAUSED\\",
    PARTIALLY_COMPLETE = \\"PARTIALLY_COMPLETE\\"
}

export enum EnumDefStatus {
    BLUE = \\"BLUE\\",
    RED = \\"RED\\",
    GREEN = \\"GREEN\\",
    ORANGE = \\"ORANGE\\",
    YELLOW = \\"YELLOW\\"
}

export interface NotUser<MetadataType = object> {
    id: string;
    metadata: MetadataType;
    def_status: EnumDefStatus | null;
    created_at: Date;
    deleted_at: Date | null;
}

export interface SomeData_2<AnObjectType = object> {
    a_number: number | null;
    a_letter: string | null;
    an_object: AnObjectType | null;
}"
`;
